Regional indicator symbols (emoji flags) broken on Windows

### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

The symbols of the regional indicators are not displayed in zed. In other IDEs, characters like "🇩🇪" turn into emoji or are displayed simply as text. But in zed there is just an invisible string. Changing fonts didn't help.

### Environment

WindowsBuildLabEx                                       : 22621.1.amd64fre.ni_release.220506-1250
WindowsProductName                                      : Windows 10 Home Single Language
CsProcessors                                            : {AMD Ryzen 9 7950X 16-Core Processor            }
PowerPlatformRole                                       : Desktop


### If applicable, add mockups / screenshots to help explain present your vision of the feature


https://github.com/user-attachments/assets/d92af7e7-993f-46bf-86b4-3fb0d1834108



### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary><pre>
<!-- Click below this line and paste or drag-and-drop your log-->
colored glyphs. (0x8898500C)
2024-09-18T21:42:45.0868679+03:00 [ERROR] The font or sequence of glyphs does not contain colored glyphs. (0x8898500C)
2024-09-18T21:42:45.0869971+03:00 [ERROR] The font or sequence of glyphs does not contain colored glyphs. (0x8898500C)
2024-09-18T21:42:45.0871325+03:00 [ERROR] The font or sequence of glyphs does not contain colored glyphs. (0x8898500C)
2024-09-18T21:42:45.0872584+03:00 [ERROR] The font or sequence of glyphs does not contain colored glyphs. (0x8898500C)
2024-09-18T21:42:45.0873769+03:00 [ERROR] The font or sequence of glyphs does not contain colored glyphs. (0x8898500C)
2024-09-18T21:42:45.0874821+03:00 [ERROR] The font or sequence of glyphs does not contain colored glyphs. (0x8898500C)
2024-09-18T21:42:45.0875572+03:00 [ERROR] The font or sequence of glyphs does not contain colored glyphs. (0x8898500C)
2024-09-18T21:42:45.0876533+03:00 [ERROR] The font or sequence of glyphs does not contain colored glyphs. (0x8898500C)
2024-09-18T21:42:45.0877576+03:00 [ERROR] The font or sequence of glyphs does not contain colored glyphs. (0x8898500C)
<!-- Click above this line and paste or drag-and-drop your log--></pre></details>

--- a/crates/gpui/src/platform/windows/direct_write.rs
+++ b/crates/gpui/src/platform/windows/direct_write.rs
@@ -813,7 +813,7 @@ impl DirectWriteState {
 
             if params.is_emoji {
                 // WARN: only DWRITE_GLYPH_IMAGE_FORMATS_COLR has been tested
-                if let Ok(enumerator) = self.components.factory.TranslateColorGlyphRun(
+                let enumerator = self.components.factory.TranslateColorGlyphRun(
                     baseline_origin,
                     &glyph_run as _,
                     None,
@@ -825,50 +825,41 @@ impl DirectWriteState {
                     DWRITE_MEASURING_MODE_NATURAL,
                     None,
                     0,
-                ) {
-                    while enumerator.MoveNext().is_ok() {
-                        let Ok(color_glyph) = enumerator.GetCurrentRun() else {
-                            break;
-                        };
-                        let color_glyph = &*color_glyph;
-                        let brush_color = translate_color(&color_glyph.Base.runColor);
-                        brush.SetColor(&brush_color);
-                        match color_glyph.glyphImageFormat {
-                            DWRITE_GLYPH_IMAGE_FORMATS_PNG
-                            | DWRITE_GLYPH_IMAGE_FORMATS_JPEG
-                            | DWRITE_GLYPH_IMAGE_FORMATS_PREMULTIPLIED_B8G8R8A8 => render_target
-                                .DrawColorBitmapGlyphRun(
-                                    color_glyph.glyphImageFormat,
-                                    baseline_origin,
-                                    &color_glyph.Base.glyphRun,
-                                    color_glyph.measuringMode,
-                                    D2D1_COLOR_BITMAP_GLYPH_SNAP_OPTION_DEFAULT,
-                                ),
-                            DWRITE_GLYPH_IMAGE_FORMATS_SVG => render_target.DrawSvgGlyphRun(
+                )?;
+                while enumerator.MoveNext().is_ok() {
+                    let Ok(color_glyph) = enumerator.GetCurrentRun() else {
+                        break;
+                    };
+                    let color_glyph = &*color_glyph;
+                    let brush_color = translate_color(&color_glyph.Base.runColor);
+                    brush.SetColor(&brush_color);
+                    match color_glyph.glyphImageFormat {
+                        DWRITE_GLYPH_IMAGE_FORMATS_PNG
+                        | DWRITE_GLYPH_IMAGE_FORMATS_JPEG
+                        | DWRITE_GLYPH_IMAGE_FORMATS_PREMULTIPLIED_B8G8R8A8 => render_target
+                            .DrawColorBitmapGlyphRun(
+                                color_glyph.glyphImageFormat,
                                 baseline_origin,
                                 &color_glyph.Base.glyphRun,
-                                &brush,
-                                None,
-                                color_glyph.Base.paletteIndex as u32,
                                 color_glyph.measuringMode,
+                                D2D1_COLOR_BITMAP_GLYPH_SNAP_OPTION_DEFAULT,
                             ),
-                            _ => render_target.DrawGlyphRun(
-                                baseline_origin,
-                                &color_glyph.Base.glyphRun,
-                                Some(color_glyph.Base.glyphRunDescription as *const _),
-                                &brush,
-                                color_glyph.measuringMode,
-                            ),
-                        }
+                        DWRITE_GLYPH_IMAGE_FORMATS_SVG => render_target.DrawSvgGlyphRun(
+                            baseline_origin,
+                            &color_glyph.Base.glyphRun,
+                            &brush,
+                            None,
+                            color_glyph.Base.paletteIndex as u32,
+                            color_glyph.measuringMode,
+                        ),
+                        _ => render_target.DrawGlyphRun(
+                            baseline_origin,
+                            &color_glyph.Base.glyphRun,
+                            Some(color_glyph.Base.glyphRunDescription as *const _),
+                            &brush,
+                            color_glyph.measuringMode,
+                        ),
                     }
-                } else {
-                    render_target.DrawGlyphRun(
-                        baseline_origin,
-                        &glyph_run,
-                        None,
-                        &brush,
-                        DWRITE_MEASURING_MODE_NATURAL,
-                    );
                 }
             } else {
                 render_target.DrawGlyphRun(
@@ -1072,7 +1063,7 @@ impl IDWriteTextRenderer_Impl for TextRenderer_Impl {
             // This `cast()` action here should never fail since we are running on Win10+, and
             // `IDWriteFontFace3` requires Win10
             let font_face = &font_face.cast::<IDWriteFontFace3>().unwrap();
-            let Some((font_identifier, font_struct, is_emoji)) =
+            let Some((font_identifier, font_struct, color_font)) =
                 get_font_identifier_and_font_struct(font_face, &self.locale)
             else {
                 return Ok(());
@@ -1093,6 +1084,13 @@ impl IDWriteTextRenderer_Impl for TextRenderer_Impl {
                 context
                     .index_converter
                     .advance_to_utf16_ix(context.utf16_index);
+                let is_emoji = {
+                    if color_font {
+                        is_color_glyph(font_face, id, color_font)
+                    } else {
+                        color_font
+                    }
+                };
                 glyphs.push(ShapedGlyph {
                     id,
                     position: point(px(context.width), px(0.0)),
@@ -1452,6 +1450,21 @@ fn get_render_target_property(
         dpiY: 96.0,
         usage: D2D1_RENDER_TARGET_USAGE_NONE,
         minLevel: D2D1_FEATURE_LEVEL_DEFAULT,
+    }
+}
+
+fn is_color_glyph(font_face: &IDWriteFontFace3, glyph_id: GlyphId, color_font: bool) -> bool {
+    unsafe {
+        let Ok(face) = font_face.cast::<IDWriteFontFace4>() else {
+            return color_font;
+        };
+        let Some(format) = face
+            .GetGlyphImageFormats(glyph_id.0 as u16, 0, u32::MAX)
+            .log_err()
+        else {
+            return color_font;
+        };
+        format != DWRITE_GLYPH_IMAGE_FORMATS_NONE
     }
 }
 

--- a/crates/gpui/src/platform/windows/direct_write.rs
+++ b/crates/gpui/src/platform/windows/direct_write.rs
@@ -1084,7 +1084,8 @@ impl IDWriteTextRenderer_Impl for TextRenderer_Impl {
                 context
                     .index_converter
                     .advance_to_utf16_ix(context.utf16_index);
-                let is_emoji = color_font && is_color_glyph(font_face, id, color_font);
+                let is_emoji = color_font
+                    && is_color_glyph(font_face, id, &context.text_system.components.factory);
                 glyphs.push(ShapedGlyph {
                     id,
                     position: point(px(context.width), px(0.0)),
@@ -1447,17 +1448,42 @@ fn get_render_target_property(
     }
 }
 
-fn is_color_glyph(font_face: &IDWriteFontFace3, glyph_id: GlyphId, color_font: bool) -> bool {
-    let Ok(face) = font_face.cast::<IDWriteFontFace4>() else {
-        return color_font;
+// One would think that with newer DirectWrite method: IDWriteFontFace4::GetGlyphImageFormats
+// but that doesn't seem to work for some glyphs, say ❤
+fn is_color_glyph(
+    font_face: &IDWriteFontFace3,
+    glyph_id: GlyphId,
+    factory: &IDWriteFactory5,
+) -> bool {
+    let glyph_run = DWRITE_GLYPH_RUN {
+        fontFace: unsafe { std::mem::transmute_copy(font_face) },
+        fontEmSize: 14.0,
+        glyphCount: 1,
+        glyphIndices: &(glyph_id.0 as u16),
+        glyphAdvances: &0.0,
+        glyphOffsets: &DWRITE_GLYPH_OFFSET {
+            advanceOffset: 0.0,
+            ascenderOffset: 0.0,
+        },
+        isSideways: BOOL(0),
+        bidiLevel: 0,
     };
-    let Some(format) = (unsafe {
-        face.GetGlyphImageFormats(glyph_id.0 as u16, 0, u32::MAX)
-            .log_err()
-    }) else {
-        return color_font;
-    };
-    format != DWRITE_GLYPH_IMAGE_FORMATS_NONE
+    unsafe {
+        factory.TranslateColorGlyphRun(
+            D2D_POINT_2F::default(),
+            &glyph_run as _,
+            None,
+            DWRITE_GLYPH_IMAGE_FORMATS_COLR
+                | DWRITE_GLYPH_IMAGE_FORMATS_SVG
+                | DWRITE_GLYPH_IMAGE_FORMATS_PNG
+                | DWRITE_GLYPH_IMAGE_FORMATS_JPEG
+                | DWRITE_GLYPH_IMAGE_FORMATS_PREMULTIPLIED_B8G8R8A8,
+            DWRITE_MEASURING_MODE_NATURAL,
+            None,
+            0,
+        )
+    }
+    .is_ok()
 }
 
 const DEFAULT_LOCALE_NAME: PCWSTR = windows::core::w!("en-US");

--- a/crates/gpui/src/platform/windows/direct_write.rs
+++ b/crates/gpui/src/platform/windows/direct_write.rs
@@ -1084,13 +1084,7 @@ impl IDWriteTextRenderer_Impl for TextRenderer_Impl {
                 context
                     .index_converter
                     .advance_to_utf16_ix(context.utf16_index);
-                let is_emoji = {
-                    if color_font {
-                        is_color_glyph(font_face, id, color_font)
-                    } else {
-                        color_font
-                    }
-                };
+                let is_emoji = color_font && is_color_glyph(font_face, id, color_font);
                 glyphs.push(ShapedGlyph {
                     id,
                     position: point(px(context.width), px(0.0)),

--- a/crates/gpui/src/platform/windows/direct_write.rs
+++ b/crates/gpui/src/platform/windows/direct_write.rs
@@ -813,7 +813,7 @@ impl DirectWriteState {
 
             if params.is_emoji {
                 // WARN: only DWRITE_GLYPH_IMAGE_FORMATS_COLR has been tested
-                let enumerator = self.components.factory.TranslateColorGlyphRun(
+                if let Ok(enumerator) = self.components.factory.TranslateColorGlyphRun(
                     baseline_origin,
                     &glyph_run as _,
                     None,
@@ -825,41 +825,50 @@ impl DirectWriteState {
                     DWRITE_MEASURING_MODE_NATURAL,
                     None,
                     0,
-                )?;
-                while enumerator.MoveNext().is_ok() {
-                    let Ok(color_glyph) = enumerator.GetCurrentRun() else {
-                        break;
-                    };
-                    let color_glyph = &*color_glyph;
-                    let brush_color = translate_color(&color_glyph.Base.runColor);
-                    brush.SetColor(&brush_color);
-                    match color_glyph.glyphImageFormat {
-                        DWRITE_GLYPH_IMAGE_FORMATS_PNG
-                        | DWRITE_GLYPH_IMAGE_FORMATS_JPEG
-                        | DWRITE_GLYPH_IMAGE_FORMATS_PREMULTIPLIED_B8G8R8A8 => render_target
-                            .DrawColorBitmapGlyphRun(
-                                color_glyph.glyphImageFormat,
+                ) {
+                    while enumerator.MoveNext().is_ok() {
+                        let Ok(color_glyph) = enumerator.GetCurrentRun() else {
+                            break;
+                        };
+                        let color_glyph = &*color_glyph;
+                        let brush_color = translate_color(&color_glyph.Base.runColor);
+                        brush.SetColor(&brush_color);
+                        match color_glyph.glyphImageFormat {
+                            DWRITE_GLYPH_IMAGE_FORMATS_PNG
+                            | DWRITE_GLYPH_IMAGE_FORMATS_JPEG
+                            | DWRITE_GLYPH_IMAGE_FORMATS_PREMULTIPLIED_B8G8R8A8 => render_target
+                                .DrawColorBitmapGlyphRun(
+                                    color_glyph.glyphImageFormat,
+                                    baseline_origin,
+                                    &color_glyph.Base.glyphRun,
+                                    color_glyph.measuringMode,
+                                    D2D1_COLOR_BITMAP_GLYPH_SNAP_OPTION_DEFAULT,
+                                ),
+                            DWRITE_GLYPH_IMAGE_FORMATS_SVG => render_target.DrawSvgGlyphRun(
                                 baseline_origin,
                                 &color_glyph.Base.glyphRun,
+                                &brush,
+                                None,
+                                color_glyph.Base.paletteIndex as u32,
                                 color_glyph.measuringMode,
-                                D2D1_COLOR_BITMAP_GLYPH_SNAP_OPTION_DEFAULT,
                             ),
-                        DWRITE_GLYPH_IMAGE_FORMATS_SVG => render_target.DrawSvgGlyphRun(
-                            baseline_origin,
-                            &color_glyph.Base.glyphRun,
-                            &brush,
-                            None,
-                            color_glyph.Base.paletteIndex as u32,
-                            color_glyph.measuringMode,
-                        ),
-                        _ => render_target.DrawGlyphRun(
-                            baseline_origin,
-                            &color_glyph.Base.glyphRun,
-                            Some(color_glyph.Base.glyphRunDescription as *const _),
-                            &brush,
-                            color_glyph.measuringMode,
-                        ),
+                            _ => render_target.DrawGlyphRun(
+                                baseline_origin,
+                                &color_glyph.Base.glyphRun,
+                                Some(color_glyph.Base.glyphRunDescription as *const _),
+                                &brush,
+                                color_glyph.measuringMode,
+                            ),
+                        }
                     }
+                } else {
+                    render_target.DrawGlyphRun(
+                        baseline_origin,
+                        &glyph_run,
+                        None,
+                        &brush,
+                        DWRITE_MEASURING_MODE_NATURAL,
+                    );
                 }
             } else {
                 render_target.DrawGlyphRun(

--- a/crates/gpui/src/platform/windows/direct_write.rs
+++ b/crates/gpui/src/platform/windows/direct_write.rs
@@ -1448,18 +1448,16 @@ fn get_render_target_property(
 }
 
 fn is_color_glyph(font_face: &IDWriteFontFace3, glyph_id: GlyphId, color_font: bool) -> bool {
-    unsafe {
-        let Ok(face) = font_face.cast::<IDWriteFontFace4>() else {
-            return color_font;
-        };
-        let Some(format) = face
-            .GetGlyphImageFormats(glyph_id.0 as u16, 0, u32::MAX)
+    let Ok(face) = font_face.cast::<IDWriteFontFace4>() else {
+        return color_font;
+    };
+    let Some(format) = (unsafe {
+        face.GetGlyphImageFormats(glyph_id.0 as u16, 0, u32::MAX)
             .log_err()
-        else {
-            return color_font;
-        };
-        format != DWRITE_GLYPH_IMAGE_FORMATS_NONE
-    }
+    }) else {
+        return color_font;
+    };
+    format != DWRITE_GLYPH_IMAGE_FORMATS_NONE
 }
 
 const DEFAULT_LOCALE_NAME: PCWSTR = windows::core::w!("en-US");

